### PR TITLE
feat(paladin): add paladin API client

### PR DESCRIPTION
**Pull request by dhaus67**
## Description

This PR can be seen as a prefactor which adds the Paladin Cloud API client in perparation to interact with the API to obtain discovered clusters.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

- See added unit tests.

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

**Comment by dhaus67** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #9774** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/9774?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/9774?utm_source=stack-comment).

File: pkg/cloudsources/paladin/client.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	token         string
+	acsVersion    string
+}
+
+func (p *paladinTransportWrapper) RoundTrip(req *http.Request) (*http.Response, error) {
+	// The Paladin Cloud API expected the Authorization header in the format '"Authorization:" <TOKEN>'
+	// instead of e.g. Bearer token format.
+	req.Header.Set("Authorization", p.token)
+
+	req.Header.Set("User-Agent", "RHACS/"+p.acsVersion)
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:52]: Maybe we can reuse user agent from https://github.com/stackrox/stackrox/blob/3113becfda03619e27cd6fce38b450edf2da67d5/pkg/clientconn/useragent.go#L47

File: pkg/cloudsources/paladin/client.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			baseTransport: proxy.RoundTripper(),
+			token:         cfg.GetCredentials().GetSecret(),
+			acsVersion:    version.GetMainVersion(),
+		},
+		Timeout: time.Minute,
+	}
+
+	return &Client{
+		httpClient: httpClient,
+		endpoint:   cfg.GetPaladinCloud().GetEndpoint(),
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:69]: where it's validated?

File: pkg/cloudsources/paladin/client.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+// NewClient creates a client to interact with Paladin Cloud APIs.
+func NewClient(cfg *storage.CloudSource) *Client {
+	httpClient := &http.Client{
+		Transport: &paladinTransportWrapper{
+			baseTransport: proxy.RoundTripper(),
+			token:         cfg.GetCredentials().GetSecret(),
+			acsVersion:    version.GetMainVersion(),
+		},
+		Timeout: time.Minute,
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:64]: That's a lot of time

File: pkg/cloudsources/paladin/client.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+		httpClient: httpClient,
+		endpoint:   cfg.GetPaladinCloud().GetEndpoint(),
+	}
+}
+
+// GetAssets returns the discovered assets from Paladin Cloud.
+func (c *Client) GetAssets(ctx context.Context) (*AssetsResponse, error) {
+	var assets AssetsResponse
+
+	if err := c.sendRequest(ctx, http.MethodGet, "/v2/assets", &assets); err != nil {
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:77]: can we add filtering to assets?

File: pkg/cloudsources/paladin/client.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+		return nil, errors.Wrap(err, "retrieving assets")
+	}
+
+	return &assets, nil
+}
+
+func (c *Client) sendRequest(ctx context.Context, method string, apiPath string, response interface{}) error {
```
Comment (stehessel) [inline pkg/cloudsources/paladin/client.go:84]: Retry GET requests?

File: pkg/cloudsources/paladin/client.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+	req, err := http.NewRequestWithContext(ctx, method, path.Join(c.endpoint, apiPath), nil)
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:85]: Which one is better path or url?
```suggestion
	req, err := http.NewRequestWithContext(ctx, method, url.JoinPath(c.endpoint, apiPath), nil)
```
https://pkg.go.dev/net/url#JoinPath

File: pkg/cloudsources/paladin/client.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	Name   string `json:"name"`
+	Type   string `json:"type"`
+	Source string `json:"source"`
+	Region string `json:"region"`
+}
+
+// Client can be used to interact with the Paladin Cloud API.
+type Client struct {
+	httpClient *http.Client
+	endpoint   string
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:37]: endpoint is in fact the base url of API

File: pkg/cloudsources/paladin/client.go [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+	if resp.Body == nil {
+		return fmt.Errorf("request failed with status %d", resp.StatusCode)
+	}
+
+	buf := &bytes.Buffer{}
+	if _, err := buf.ReadFrom(resp.Body); err != nil {
+		return fmt.Errorf("request failed with status %d", resp.StatusCode)
+	}
+
+	return fmt.Errorf("request failed with status %d and response %s", resp.StatusCode, buf.String())
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:118]:
```suggestion
	return fmt.Errorf("request failed with status %d and response %q", resp.StatusCode, buf.String())
```

File: pkg/cloudsources/paladin/client.go [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+	if err := json.NewDecoder(resp.Body).Decode(response); err != nil {
+		return errors.Wrap(err, "decoding response body")
+	}
+
+	return nil
+}
+
+func getErrorResponse(resp *http.Response) error {
+	if resp.Body == nil {
+		return fmt.Errorf("request failed with status %d", resp.StatusCode)
```
Comment (janisz) [inline pkg/cloudsources/paladin/client.go:110]:
```suggestion
		return httputil.NewError(resp.StatusCode, "request failed")
```

File: pkg/cloudsources/paladin/client.go (hunk 2 above)
Comment (dhaus67) [reply pkg/cloudsources/paladin/client.go]: Validated it now within the function + ensured that default scheme is HTTPS.

Additionally validation is done in the datastore to ensure that the endpoint is a non-local one.

File: pkg/cloudsources/paladin/client.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+	Assets []Asset `json:"assets,omitempty"`
+}
+
+// Asset holds the asset as returned by the Paladin Cloud API.
+type Asset struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	Type   string `json:"type"`
+	Source string `json:"source"`
+	Region string `json:"region"`
```
Comment (stehessel) [inline pkg/cloudsources/paladin/client.go:35]: `FirstDiscoveryDate` is missing.

File: pkg/cloudsources/paladin/client.go [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+}
+
+// Asset holds the asset as returned by the Paladin Cloud API.
+type Asset struct {
+	ID               string    `json:"id"`
+	Name             string    `json:"name"`
+	Type             string    `json:"type"`
+	Source           string    `json:"source"`
+	Region           string    `json:"region"`
+	FirsDiscoveredAt time.Time `json:"firstDiscoveryDate"`
```
Comment (stehessel) [inline pkg/cloudsources/paladin/client.go:41]:
```suggestion
	FirstDiscoveredAt time.Time `json:"firstDiscoveryDate"`
```

**Review by stehessel — APPROVED** [review]

**Pull request merged**